### PR TITLE
[aml]Register lirc in AML after #13761

### PR DESCRIPTION
**Pull request by gelotus**
See title,

Sigsegv on
void CLirc::Process()
{
  m_profileId = CServiceBroker::GetProfileManager().GetCurrentProfileId();
  m_irTranslator.Load("Lircmap.xml");

In LIRC.cpp

### CODE REVIEW

**Comment by adamg88** [comment]
I already tried this, doesn't work, causes seg fault upon startup.

**Comment by fritsch** [comment]
So then: post the backtrace, you seem to have the hardware, most of us don't - that's why this code part was not part of the original PR.

**Comment by gelotus** [comment]
Kodi stack trace:
```
################################################################################
# ... output of cat /storage/.kodi/temp/kodi_crashlog_20180417104451.log
# LibreELEC release: S912.arm
# 
################################################################################
############## kodi CRASH LOG ###############

################ SYSTEM INFO ################
 Date: Tue Apr 17 10:44:51 UTC 2018
 kodi Options: --standalone -fs
 Arch: aarch64
 Kernel: Linux 3.14.29 #1 SMP Tue Apr 17 02:01:00 CEST 2018
############## END SYSTEM INFO ##############

############### STACK TRACE #################
=====>  Core file: /storage/.cache/cores/core.!usr!lib!kodi!kodi.bin.1523961889.3074
        =========================================
[New LWP 3077]
[New LWP 3074]
Core was generated by `/usr/lib/kodi/kodi.bin --standalone -fs'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x0077f910 in CServiceManager::GetProfileManager() ()
[Current thread is 1 (LWP 3077)]

Thread 2 (LWP 3074):
#0  0x002fca80 in ?? ()
#1  0x002fca7c in ?? ()
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 1 (LWP 3077):
#0  0x0077f910 in CServiceManager::GetProfileManager() ()
#1  0x007fcba0 in CLirc::Process() ()
#2  0x008ff860 in CThread::Action() ()
#3  0x0092d200 in CThread::staticThread(void*) ()
#4  0xf7533d40 in ?? ()
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
############# END STACK TRACE ###############
```

**Comment by gelotus** [comment]
Syslog:
```
[   23.102216@0] codec:vdec_create instance ffffff8004025000, total 1
[   23.134245@0] Lirc[2951]: unhandled level 3 translation fault (11) at 0x0000006c, esr 0x92000007
[   23.134256@0] pgd = ffffffc0618f3000
[   23.134260@0] [0000006c] *pgd=000000006194a003, *pmd=000000006185d003, *pte=0000000000000000
[   23.134266@0] 
[   23.134274@0] CPU: 0 PID: 2951 Comm: Lirc Tainted: G           O 3.14.29 #1
[   23.134278@0] task: ffffffc061cdca80 ti: ffffffc0631ac000 task.ti: ffffffc0631ac000
[   23.134295@0] PC is at 0x77fa30
[   23.134299@0] LR is at 0x7fccc0
[   23.134302@0] pc : [<000000000077fa30>] lr : [<00000000007fccc0>] pstate: 20000010
[   23.134306@0] sp : 00000000efaffcc8
[   23.134309@0] x12: 0000000000000b87 
[   23.134312@0] x11: 00000000efaffe7c x10: 000000000092d24c 
[   23.134316@0] x9 : 00000000efb00340 x8 : 000000000120d66c 
[   23.134321@0] x7 : 00000000efaffd20 x6 : 00000000efaffd18 
[   23.134325@0] x5 : 00000000efaffd28 x4 : 000000000120d460 
[   23.134329@0] x3 : 00000000011a4860 x2 : 00000000fffffcdc 
[   23.134335@0] x1 : 0000000000000000 x0 : 0000000000000000 
```

**Comment by gelotus** [comment]
The strange thing is if I remove this:
```
--- a/xbmc/platform/linux/input/LIRC.cpp
+++ b/xbmc/platform/linux/input/LIRC.cpp
@@ -52,44 +52,6 @@
 {
   m_profileId = CServiceBroker::GetProfileManager().GetCurrentProfileId();
   m_irTranslator.Load("Lircmap.xml");
-
-  while (!m_bStop)
-  {
-    {
-      CSingleLock lock(m_critSection);
-      m_fd = lirc_init(const_cast<char*>("kodi"), 0);
-      if (m_fd <= 0)
-      {
-        CSingleExit lock(m_critSection);
-        Sleep(1000);
-        continue;
-      }
-    }
-
-    char *code;
-    while (!m_bStop)
-    {
-      int ret = lirc_nextcode(&code);
-      if (ret < 0)
-      {
-        lirc_deinit();
-        Sleep(1000);
-        break;
-      }
-      if (code != nullptr)
-      {
-        if (m_profileId != CServiceBroker::GetProfileManager().GetCurrentProfileId())
-        {
-          m_profileId = CServiceBroker::GetProfileManager().GetCurrentProfileId();
-          m_irTranslator.Load("Lircmap.xml");
-        }
-        ProcessCode(code);
-        free(code);
-      }
-    }
-  }
-
-  lirc_deinit();
 }

 void CLirc::ProcessCode(char *buf)
```
for testing purpose, it also fail with the same error

**Comment by gelotus** [comment]
i'm also thinking a strange race condition, i can try

**Comment by fritsch** [comment]
yeah, save the return value of InitWindow system, afterwards RegisterLirc and return the temporary. It's a workaround for a deeper issue - but let's first try to come into the error's region :-)

**Comment by FernetMenta** [comment]
CServiceBroker::GetProfileManager() needs to return a pointer than can be checked for nullptr by clients

**Comment by fritsch** [comment]
What you called "Fix" is not a fix, but a workaround :-)

**Comment by FernetMenta** [comment]
a workaround that would fail soon when order of init will be changed again

**Comment by gelotus** [comment]
ok whatever it is now works, good hint :)

**Comment by gelotus** [comment]
@FernetMenta you mean to check against nullptr and if it so sleep the thread?

**Comment by fritsch** [comment]
You can currently not check vs. nullptr in your code - the issue is how ServiceManager manages ProfileManager (and others). That one cannot decide already.

Btw. with your fixup you reset the ptr two times now.

**Comment by gelotus** [comment]
@fritsch no, i'm talking about a hypothethical better implementation, but if there is a wip in ServiceManager better to leave so?

**Comment by fritsch** [comment]
I'd say squash your commits (keep the one reset that has it working). The later implementation / fix does not affect you, cause Lirc will check for the nullptr when it is registered.

**Comment by FernetMenta** [comment]
nope, we fix it properly. otherwise folks start crying again when I change the order of init

**Comment by FernetMenta** [comment]
> you mean to check against nullptr and if it so sleep the thread?

no need to sleep. profile id is only used to test for changes. if we don't get an id, there won't be a compare

**Comment by fritsch** [comment]
Not asking for not proper fixing it - but keeping AML platform broken for the time being is not good and will just put private patches into LE and other distributions. Also the proper fix is not within AML code.

**Comment by FernetMenta** [comment]
it is only broken for a very small group using lirc. most remotes are identified as keyboards and don't require lirc. we don't keep lirc unfunctional for weeks but just for a few days.
better lirc does not work for the time being than aml would crash for all.

File: xbmc/windowing/amlogic/WinSystemAmlogic.cpp [hunk 1]
```diff
@@ -101,6 +101,9 @@ bool CWinSystemAmlogic::InitWindowSystem()
   CRendererAML::Register();
 
   aml_set_framebuffer_resolution(1920, 1080, m_framebuffer_name);
+  
+  m_lirc.reset(OPTIONALS::LircRegister()); 
+  CLog::Log(LOGDEBUG, "OPTIONALS::LircRegister -- register Lirc client");
```
Comment (lrusak) [inline xbmc/windowing/amlogic/WinSystemAmlogic.cpp]: I don't think this log lines is needed

**Pull request merged**